File: Java8-JavaFX/src/sevenguis/circledrawer/undo/UndoManager.java
package sevenguis.circledrawer.undo;

import java.util.Stack;

public class UndoManager {

    Stack<UndoableEdit> history;
    int cursor;

    public UndoManager() {
        history = new Stack<>();
        cursor = -1;
    }

    public void addEdit(UndoableEdit edit) {
        int toPop = history.size()-1 - cursor;
        while (toPop != 0) {
            history.pop();
            toPop--;
        }
        history.push(edit);
        cursor++;
    }

    public void undo() {
        if (cursor >= 0) {
            history.get(cursor).undo();
            cursor--;
        }
    }

    public void redo() {
        if (cursor < history.size()-1) {
            cursor++;
            history.get(cursor).redo();
        }
    }

}

File: Qt5/circledrawer/source/undo/undomanager.cpp
#include "undomanager.h"

UndoManager::UndoManager()
{
    history = new QStack<UndoableEdit *>();
    cursor = -1;
}

void UndoManager::addEdit(UndoableEdit *edit) {
    int toPop = history->size() - 1 - cursor;

    while (toPop != 0)
    {
        history->pop();
        toPop--;
    }

    history->push(edit);
    cursor++;
}

void UndoManager::undo()
{
    if (cursor >= 0)
    {
        history->at(cursor)->undo();
        cursor--;
    }
}

void UndoManager::redo() {
    if (cursor < history->size()-1)
    {
        cursor++;
        history->at(cursor)->redo();
    }
}

UndoManager::~UndoManager()
{

}
